src/ucs/sys/event_set.h

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (C) Hiroyuki Sato. 2019. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_EVENT_SET_H
+#define UCS_EVENT_SET_H
+
+#include <ucs/type/status.h>
+
+/**
+ * ucs_sys_event_set structure used in ucs_event_set_XXX functions.
+ *
+ */
+typedef struct ucs_sys_event_set {
+    int                epfd;
+} ucs_sys_event_set_t;
+
+typedef ucs_status_t (*event_set_handler_t)(int nready, void *event);
+
+/**
+ * ucs_event_set_type_t member is a bit set composed using the following
+ * available event types
+ */
+typedef enum {
+    UCS_EVENT_SET_EVREAD  = 1,
+    UCS_EVENT_SET_EVWRITE = 1 << 1,
+    UCS_EVENT_SET_EVNONE =  1 << 2
+} ucs_event_set_type_t;
+
+/**
+ * Allocate ucs_sys_event_set_t structure.
+ *
+ * @param [out] event_set_p  Event set pointer to initialize.
+ *
+ * @return UCS_OK on success or an error code on failure.
+ */
+ucs_status_t ucs_event_set_create(ucs_sys_event_set_t **event_set_p);
+
+/**
+ * Register the target event.
+ *
+ * @param [in] event_set_p  Event set pointer to initialize.
+ * @param [in] event_fd     Register the target file descriptor fd.
+ * @param [in] events       Operation events.
+ * @param [in] udata        User data.
+ *
+ * @return UCS_OK on success or an error code on failure.
+ */
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata);
+
+/**
+ * Modify the target event.
+ *
+ * @param [in] event_set    Event set created by ucs_event_set_create.
+ * @param [in] event_fd     Register the target file descriptor fd.
+ * @param [in] events       Operation events.
+ * @param [in] udata        User data.
+ *
+ * @return UCS_OK on success or an error code on failure.
+ */
+ucs_status_t ucs_event_set_mod(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata);
+
+/**
+ * Remove the target event.
+ *
+ * @param [in] event_set    Event set created by ucs_event_set_create.
+ * @param [in] event_fd     Register the target file descriptor fd.
+ *
+ * @return UCS_OK on success or an error code on failure.
+ */
+ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd);
+
+/**
+ * Wait for an I/O events
+ *
+ * @param [in] event_set    Event set created by ucs_event_set_create.
+ * @param [in] event_fd     Register the target file descriptor fd.
+ *
+ * @return UCS_OK on success or an error code on failure.
+ */
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
```

Comment:
need to pass used-defined argument to the callback

Reply:
how about passing argument to the callback itself which may depend on the context where the callback is called? for example, pass tcp iface pointer in it

Reply:
I misunderstood epoll_wait usage. changed callback functions.